File: FSharp/Library/Math.fs
namespace Library

module Math =
    let rec Factorial n =
        match n with
        | 0 -> 1
        | _ -> n * Factorial (n - 1)

    let rec FactorialTail n =
        let rec accumulatedFactorial x acc =
            match x with
            | 0 -> acc
            | _ -> accumulatedFactorial (x - 1) (x * acc)   

        accumulatedFactorial n 1                 

    let FactorialIterative n =
        let mutable result = 1
        for i = n downto 1 do
            result <- result * i
        result

File: CSharp/Library/Math.cs
namespace Library
{
    public static class Math
    {
        public static int FactorialIterative(int n)
        {
            var result = 1;

            for (var i = n; i > 1; i--)
                result *= i;

            return result;
        }

        public static int Factorial(int n) =>
            n <= 1 
            ? 1 
            : Factorial(n - 1) * n;

        public static int FactorialTail(int n) {
            int accumulatedFactorial(int n, int acc) =>
                n <= 1 
                ? acc 
                : accumulatedFactorial(n - 1, acc * n);
            
            return accumulatedFactorial(n, 1);
        }
    }
}